### PR TITLE
feat: add `try_parse_json` function

### PR DESCRIPTION
**Pull request by davidlghellin**
part https://github.com/lakehq/sail/issues/862

### CODE REVIEW

File: crates/sail-function/src/scalar/variant/spark_try_parse_json.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+use crate::scalar::variant::utils::helper::{try_field_as_string, try_parse_string_scalar};
+
+/// Returns a Variant from a JSON string, or NULL if the JSON is invalid.
+/// This is the safe version of `parse_json`.
+///
+/// Like Spark, trailing content after valid JSON is ignored (the valid prefix is parsed).
+///
+/// <https://spark.apache.org/docs/latest/api/sql/index.html#try_parse_json>
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct SparkTryParseJsonUdf {
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply crates/sail-function/src/scalar/variant/spark_try_parse_json.rs]: Covered by 29 feature tests in try_parse_json.feature — including NULL input, invalid JSON, trailing garbage, multi-row, unicode, and edge cases. All validated against Spark JVM.

File: crates/sail-function/src/scalar/variant/spark_try_parse_json.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+/// Try to append a JSON string to the builder. Returns true if successful.
+fn try_append_json(builder: &mut VariantArrayBuilder, json_str: &str) -> bool {
+    match try_parse_json_lenient(json_str) {
+        Some(value) => {
+            let json_str = value.to_string();
+            builder.append_json(json_str.as_str()).is_ok()
+        }
+        None => false,
+    }
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply crates/sail-function/src/scalar/variant/spark_try_parse_json.rs]: Fixed — fast path added: builder.append_json(json_str) is tried first; the lenient serde_json fallback only runs for inputs with trailing content.

[the remark this answers is not in the document]
Comment (davidlghellin) [reply crates/sail-function/src/scalar/variant/spark_try_parse_json.rs]: Reverted — the fast path bypasses StrictValue duplicate-key rejection, causing try_parse_json('{"a":1,"a":2}') to return non-NULL instead of NULL (diverging from Spark). The serde_json path must run first to preserve that semantic.

https://github.com/lakehq/sail/pull/1636/changes/b7c09c3fe6e69813dda4b36f09449db61b218ffb

File: crates/sail-function/src/scalar/variant/spark_parse_json.rs [hunk 3]
```diff
@@ -1,49 +1,178 @@
...
+    let mut stream = serde_json::Deserializer::from_str(json_str).into_iter::<StrictValue>();
+    stream.next().and_then(|r| r.ok()).map(|sv| sv.0)
+}
+
+/// Try to append a JSON string to the builder leniently. Returns true if successful.
+fn try_append_json(builder: &mut VariantArrayBuilder, json_str: &str) -> bool {
+    // Must go through try_parse_json_lenient (StrictValue) to reject duplicate keys,
+    // matching Spark's try_parse_json semantics (returns NULL on duplicate keys).
+    match try_parse_json_lenient(json_str) {
+        Some(value) => builder.append_json(value.to_string().as_str()).is_ok(),
```
Comment (linhr) [inline crates/sail-function/src/scalar/variant/spark_parse_json.rs]: `builder.append_json(value.to_string().as_str())` would have unnecessary JSON parsing and serialization, but `value` is already of type `Value`.

This function seems more suitable here:

```rust
pub fn append_json(json: &Value, builder: &mut impl VariantBuilderExt) -> Result<(), ArrowError>;
```

File: crates/sail-function/src/scalar/variant/spark_parse_json.rs [hunk 4]
```diff
@@ -1,49 +1,178 @@
...
+        "[MALFORMED_RECORD_IN_PARSING.WITHOUT_SUGGESTION] Malformed records are detected in record parsing: {record}."
+    )
+}
+
+/// Strict-path append: accept trailing garbage (first valid prefix), error
+/// with `MALFORMED_RECORD_IN_PARSING` on unparseable input.
+fn append_json_strict(builder: &mut VariantArrayBuilder, json_str: &str) -> Result<()> {
+    match try_parse_json_lenient(json_str) {
+        Some(value) => builder
+            .append_json(value.to_string().as_str())
```
Comment (linhr) [inline crates/sail-function/src/scalar/variant/spark_parse_json.rs]: Same here for performance concerns.

**Review by linhr — COMMENTED** [review]
Overall this looks good!

We can merge this after addressing the comments regarding performance.

**Review by linhr — APPROVED** [review]
Great! Thanks!
